addons/mail/static/src/new/thread/thread_service.js

```diff
@@ -601,6 +608,68 @@ export class ThreadService {
         }
         return message;
     }
+
+    /**
+     * @param {Thread} thread
+     */
+    localMessageUnreadCounter(thread) {
+        // By default trust the server up to the last message it used
+        // because it's not possible to do better.
+        let baseCounter = thread.serverMessageUnreadCounter;
+        let countFromId = thread.serverLastSeenMsgBySelf ? thread.serverLastSeenMsgBySelf : 0;
+        const orderedMessages = this.message
+            .sortMessages(thread)
+            .map((messageId) => this.store.messages[messageId]);
+        // But if the client knows the last seen message that the server
+        // returned (and by assumption all the messages that come after),
+        // the counter can be computed fully locally, ignoring potentially
+        // obsolete values from the server.
+        const lastSeenByCurrentPartnerMessageId = this.lastSeenByCurrentPartnerMessageId(
+            thread,
+            orderedMessages
+        );
+        const firstMessage = orderedMessages[0];
+        if (
+            firstMessage &&
+            lastSeenByCurrentPartnerMessageId &&
+            lastSeenByCurrentPartnerMessageId >= firstMessage.id
+        ) {
+            baseCounter = 0;
+            countFromId = lastSeenByCurrentPartnerMessageId;
+        }
+        // Include all the messages that are known locally but the server
+        // didn't take into account.
+        return orderedMessages.reduce((total, message) => {
+            if (message.id <= countFromId) {
+                return total;
```

Comment:
I prefer to remove all comments. It's not that useful, and it just adds noise. Code is easier to ready without them.